### PR TITLE
NEW Make use of ViewModeToggle component.

### PR DESCRIPTION
**Pull request by raissanorth**
Allows users to toggle between edit, preview, and split mode in the CMS. 

Introduces dependencies on `react-resize-aware` library and `Enzyme` testing utility. The former is used to retrieve the screen width and determine if split mode should be enabled based on it. The latter is used to shallow render `CampaignAdminList` in the unit tests, as we don't want to test the child components `ViewModeToggle` and `Preview`. These components are now provided via Injector.

Removed media queries in favour of rendering preview and/or edit based on the state stored in `ViewModeToggle`'s store.

Refactored `CampaignAdminList`'s render method, to make it more readable. Moved the logic to render the detail/edit view into its own method - `renderCampaignAdminListDetail`

### CODE REVIEW

File: client/src/containers/CampaignAdmin/CampaignAdmin.scss [hunk 1]
```diff
@@ -3,33 +3,27 @@
...
+    //  @include media-breakpoint-down(lg) {
+    //    border: 0;
+    //  }
+    //
+    //  &:hover {
+    //    background-color: $list-group-hover-bg;
+    //    color: $list-group-link-color;
+    //    border-color: $list-group-border-color;
+    //  }
+    //}
```
Comment (robbieaverill) [inline client/src/containers/CampaignAdmin/CampaignAdmin.scss]: We can remove this if it's not required any more

File: client/src/containers/CampaignAdmin/CampaignAdmin.js [hunk 2]
```diff
@@ -536,6 +545,10 @@ function mapDispatchToProps(dispatch) {
     breadcrumbsActions: bindActionCreators(breadcrumbsActions, dispatch),
     campaignActions: bindActionCreators(campaignActions, dispatch),
     recordActions: bindActionCreators(recordActions, dispatch),
+
+    onResize(panelWidth) {
```
Comment (robbieaverill) [inline client/src/containers/CampaignAdmin/CampaignAdmin.js]: Could you add a proptype for this please?

File: client/src/containers/CampaignAdmin/CampaignAdminList.js [hunk 3]
```diff
@@ -279,53 +293,114 @@ class CampaignAdminList extends Component {
...
     );
   }
 
   renderPreview(itemLinks, itemId) {
-    let preview = null;
     const previewClasses = classnames([
       'flexbox-area-grow',
       'fill-height',
       'preview',
-      'campaign-admin__campaign-preview',
```
Comment (robbieaverill) [inline client/src/containers/CampaignAdmin/CampaignAdminList.js]: Any reason this was removed?

File: client/src/containers/CampaignAdmin/CampaignAdminList.js [hunk 4]
```diff
@@ -279,53 +293,114 @@ class CampaignAdminList extends Component {
...
+      />
+    );
+  }
+
+  /**
+   * Renders the details section of the campaign list.
+   *
+   * @param body
+   *
+   * @return object
```
Comment (robbieaverill) [inline client/src/containers/CampaignAdmin/CampaignAdminList.js]: I think these JS doc annotations are a little out

File: client/src/containers/CampaignAdmin/CampaignAdminList.js [hunk 5]
```diff
@@ -279,53 +293,114 @@ class CampaignAdminList extends Component {
...
+    // Hide when the preview mode is explicitly enabled
+    if (previewState === 'preview') {
+      return null;
+    }
+
+    let classNames = 'fill-height campaign-admin__campaign-items';
+    // Modify width when the edit mode is enabled
+    if (previewState === 'edit') {
+      classNames = 'fill-height campaign-admin__campaign-items-edit';
+    }
```
Comment (robbieaverill) [inline client/src/containers/CampaignAdmin/CampaignAdminList.js]: Using the classnames lib here might be a little tidier

File: client/src/containers/CampaignAdmin/CampaignAdminList.js [hunk 6]
```diff
@@ -279,53 +293,114 @@ class CampaignAdminList extends Component {
...
+
+    return (
+      <div className={classNames} aria-expanded="true">
+        <Toolbar showBackButton onBackButtonClick={onBackButtonClick}>
+          <Breadcrumb multiline />
+        </Toolbar>
+        {newItemInfo}
+        <div className={bodyClass.join(' ')}>
```
Comment (robbieaverill) [inline client/src/containers/CampaignAdmin/CampaignAdminList.js]: Yeah again use classnames

**Comment by raissanorth** [comment]
Note that this is dependant on https://github.com/silverstripe/silverstripe-admin/pull/560

File: client/src/containers/CampaignAdmin/CampaignAdminList.js (hunk 3 above)
Comment (raissanorth) [reply client/src/containers/CampaignAdmin/CampaignAdminList.js]: This class was used to hide the preview on smaller screens. Media queries are no longer used for this behaviour, but the ViewModeToggle's state. Therefore the class is redundant.

Comment (robbieaverill) [reply client/src/containers/CampaignAdmin/CampaignAdminList.js]: Sure, but removing it is kind of a breaking change - you've already removed the styles for it, but someone else might be relying on it too

Comment (raissanorth) [reply client/src/containers/CampaignAdmin/CampaignAdminList.js]: That makes sense, and is something I haven't had considered. I will revert my change.

File: client/src/containers/CampaignAdmin/CampaignAdminList.js [hunk 7]
```diff
@@ -252,7 +253,20 @@ class CampaignAdminList extends Component {
...
+  renderViewModeToggle() {
+    const { ViewModeComponent } = this.props;
+
+    return (
+      <ViewModeComponent
+        id="view-mode-toggle-in-edit-nb"
+        area={'edit'}
+      />
+    );
+  }
```
Comment (robbieaverill) [inline client/src/containers/CampaignAdmin/CampaignAdminList.js]: It looks like this method isn't used at all - perhaps you could remove it. The component is rendered in the `render()` method directly and doesn't need to be found to the instance in construction I don't think

**Comment by flamerohr** [comment]
@robbieaverill make sure to bump the version dependency in the package.json as well

also, I'm not sure why tests are failing for `react-apollo`

**Comment by robbieaverill** [comment]
>also, I'm not sure why tests are failing for react-apollo

@flamerohr FYI this was because the admin module npm dependencies weren't installed. I reproduced this on the latest 1 branch for campaign-admin in Travis, we've pushed a fix for it which is stolen from asset-admin

**Comment by robbieaverill** [comment]
Merge on green

**Comment by flamerohr** [comment]
oh I see *cringe* lol

**Pull request merged**